cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -126,6 +132,78 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws I
                              limit);
   }
 
+  @Override
+  public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, String orderByField, SortOrder sortOrder)
+    throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Scan range {} with limit {}, sort field {} and sort order {}", schema.getTableId(),
+              keyRange, limit, orderByField, sortOrder);
+    if (!schema.isIndexColumn(orderByField)) {
+      throw new InvalidFieldException(schema.getTableId(), orderByField, "is not an indexed column");
+    }
+    ScannerIterator scannerIterator = null;
+    Comparator<StructuredRow> comparator = sortOrder.equals(SortOrder.ASC) ?
+      (row1, row2) ->
+        compareRowFields(row1, row2, orderByField)
+      : (row1, row2) ->
+      compareRowFields(row2, row1, orderByField);
+    List<StructuredRow> sorted = new ArrayList<>();
+
+    try {
+      scannerIterator = new ScannerIterator(getScanner(keyRange), schema);
+      while (scannerIterator.hasNext()) {
+        StructuredRow row = scannerIterator.next();
+        sorted.add(row);
+      }
+    } finally {
+      if (scannerIterator != null) {
+        scannerIterator.close();
+      }
+    }
+    Collections.sort(sorted, comparator);
```

Comment:
the previous implementation with the PriorityQueue was better. Putting it into an ArrayList and then sorting it involves an extra pass through the data.

Reply:
so the point of putting it in an arraylist is because the priority-queue iterator returns the elements in random order

Reply:
Yeah priority queue only returns sorted elements when you call the `poll` method (the top one is min or max), the iterator just iterates element in the heap, which is not sorted

Reply:
oh... did not know that. Sorry about the back and forth, but if that's the case then the previous AbstractIterator implementation is better, as it prevents the extra pass through the data.

Reply:
no problem, I'll change it back